### PR TITLE
Project 4: Data Streaming Simulation in LLVM

### LINKED ISSUE #140: Project 4 Proposal: Data Streaming Simulation in LLVM
**What will you do?**
Implement data streaming simulation with FIFO for LLVM JIT execution in HeteroCL. For example, we can implement with multiple threads or single thread using external libraries like pthread. We will also simulate the FIFO depth.

**How will you do it?**
1. Create new IR nodes for data streaming operations.
2. Develop an external library for data streaming simulation in LLVM JIT execution.

**How will you empirically measure success?**
We will 1) develop different test cases in HeteroCL; 2) simulate the programs with and without streaming in LLVM JIT execution; 3) compare the results. Aside from verifying the correctness, we also expect the simulation time to be shorter compared with the original simulation without data streaming.

**Team members:**
@seanlatias

### PR DESCRIPTION
**Pull request by hecmay**
Closes #140 with @seanlatias 

In this project we implemented data streaming simulation with FIFO for LLVM JIT execution in HeteroCL. The simulation flow models each modules as a separate thread and performs scheduling and before invoking the multi-thread execution.

Code available [here](https://github.com/Hecmay/heterocl)

### CODE REVIEW

File: content/blog/2019-12-18-heterocl-stream/index.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+#### Stream Buffers
+
+In this work, we implement the streams with buffers that act like FIFOs. Instead of popping from or pushing to the buffers, we maintain a **head** and a **tail** pointer for each buffer. The pointers are stored as integer numbers. The head pointer points to the next element that will be read from, and the tail pointer points to the next element that will be written to. We update the pointers each time an element is written to or read from the buffer. We need to perform modulo operations if the pointer value is greater than the buffer size (i.e., FIFO depth). Since we may have two threads updating the pointers at the same time, we declare them as atomic numbers. Finally, we maintain a map so that we can access a stream according to its ID.
```
Comment (sampsyo) [inline content/blog/2019-12-18-heterocl-stream/index.md]: > we declare them as atomic numbers

This is a little hard to interpret without more context… what exactly is an "atomic number"?

I assume the answer is that you wrote simulation framework in C++, and you used [std:atomic](https://en.cppreference.com/w/cpp/atomic/atomic) as an alternative to using mutexes to synchronize accesses to the counters.

File: content/blog/2019-12-18-heterocl-stream/index.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+We can see that without applying streaming, the production of ``D`` is not affected by ``M2``. However, if we specify ``C`` to be streamed from ``M2`` to ``M1``, the original memory read of ``C`` in ``M1`` now becomes a blocking read. This also demonstrates that without the simulation support for streaming, some hardware behaviors cannot be correctly represented.
+
+#### Realistic Benchmark
+
+We also show the evaluation results from a realistic benchmark, which is more complicated than the synthetic tests in the unit tests. Due to time limitation, we only use the Sobel edge detector, which is a popular edge detecting algorithm in image processing. We compare the results with the software simulation tool provided by the HLS compiler. More specifically, we first generate Vivado HLS code with ``hls::stream``. Then we use ``csim`` to run the software simulation. The evaluation results are shown below. We also show the time overhead due to compilation.
+
+| Simulation Method | Simulation Time (s) | Compilation Overhead (s) | Total Run Time (s) |
+|:---------:|------------|------------|--------|
+|LLVM JIT|0.00094|0|0.00094|
+|Vivado HLS csim|1.63|1.29|1.92|
```
Comment (sampsyo) [inline content/blog/2019-12-18-heterocl-stream/index.md]: Hmm, it seems like the simulation time + compilation overhead should equal the total run time. But 1.63 + 1.29 ≠ 1.92.

**Review by sampsyo — CHANGES_REQUESTED** [review]
Good work! It looks like you indeed set out to do what you wanted to. There weren't terribly many connections here to 6120-style compilers research but it clearly let you apply your LLVM skills.

File: content/blog/2019-12-18-heterocl-stream/index.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+#### Parallel Execution with Threads
+
+After we assign each module with a timestep, we can start to execute them via threads. Before we execute a module with a new thread, we check whether all modules assigned with smaller timesteps are completed. In other words, we first check whether all modules assigned with smaller timesteps are fired. If not, we schedule the current module to be executed in the future by pushing it into a sorted execution queue. Then, if all modules with smaller timesteps are fired, we check whether they are finished. If not, we perform thread synchronization (e.g., by using ``thread.join()`` in C++). Finally, we need to execute the modules in the execution queue. Since the queue is sorted, we do not need to worry about new modules being inserted into the queue.
```
[the remark this answers is not in the document]
Comment (seanlatias) [reply content/blog/2019-12-18-heterocl-stream/index.md]: You are right. A heap should be more efficient.

**Review by sampsyo — APPROVED** [review]
Looking good!

**Pull request merged**